{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "cf08c110-330f-4d6e-b73b-bae5bef9fc7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pandasgui as pg \n",
    "\n",
    "curated_vars = pd.read_csv(\"../data/row_curateddata.csv\", dtype={\"Pubmed_id\":\"object\"})\n",
    "snp_table = pd.read_csv(\"../data/july20th_previous_database/snps_July20.csv\")\n",
    "pmid_table = pd.read_csv(\"../data/july20th_previous_database/study_July20.csv\", dtype={\"reference_id\":\"object\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "4a882a58-e146-4981-9dd6-3409eb434f4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "filter_list = []\n",
    "for var in curated_vars.Variant_ID:\n",
    "    if \"*\" in var:\n",
    "        filter_list.append(True)\n",
    "    else:\n",
    "        filter_list.append(False)\n",
    "\n",
    "star_alleles  = curated_vars[filter_list]\n",
    "snp = curated_vars[np.invert(filter_list)]\n",
    "snp.drop(columns=[\"Unnamed: 0\"], inplace=True)\n",
    "snp.rename(columns={\"Variant_ID\":\"rs_id\"}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "52ca338a-4f39-4308-bbd4-2ddd18a0f367",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add genes \n",
    "genes = snp_table[[\"rs_id\", \"gene_id\" ]]\n",
    "new_snp = pd.merge(snp, genes, on=\"rs_id\").drop_duplicates()\n",
    "\n",
    "# add reference \n",
    "pmid= pmid_table[[\"ref_id\", \"reference_id\" ]]\n",
    "new_snp = pd.merge(new_snp, pmid, left_on=\"Pubmed_id\",right_on=\"reference_id\").drop_duplicates()\n",
    "\n",
    "# add the other columns \n",
    "last_merg_table = snp_table[[\"rs_id\", \"id\", \"reference_id\", \"drug_id\", \"ID_diseases\", \"source\", \"id_in_source\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "b4578e09-eafe-4588-a0f7-5a532db6a66a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<pandasgui.gui.PandasGui at 0x7fe8abefab80>"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pg.show(new_snp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "78dc6028-0c06-4f67-9d8f-3fcb89e11cee",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_merged = pd.merge(new_snp, last_merg_table, how=\"left\",\n",
    "                      left_on=[\"rs_id\", \"ref_id\"], right_on = [\"rs_id\", \"reference_id\"])\n",
    "raw_merged = raw_merged[[\"id\",\"rs_id\", \"gene_id\", \"drug_id\", \"ID_diseases\", \"phenotype\",\"reference_id_y\", \"p-value\",\n",
    "           \"source\", \"id_in_source\", \"region\", \"Ethnicity\", \"Notes\", \"origin_of_participants\"  ]].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "10494586-5ebc-44d7-a15e-fc7133373666",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_merged.to_csv(\"../data/new_curated_snps.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "57b25178-2b7d-4ab2-a44b-8b8dcf4699b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/houcem/env_module/modules/software/miniconda3/envs/jupyter_env/lib/python3.9/site-packages/pandas/core/frame.py:4441: SettingWithCopyWarning:\n",
      "\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "\n"
     ]
    }
   ],
   "source": [
    "star = pd.read_csv('../data/july20th_previous_database/star_July20.csv')\n",
    "#star_alleles.drop(columns=[\"Unnamed: 0\"], inplace=True)\n",
    "star_alleles.rename(columns={\"Variant_ID\":\"star_annotation\"}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "id": "c1b80c7e-f765-4a56-be86-a48b27a1cdf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes= star[[\"star_annotation\", \"gene_id\" ]]\n",
    "new_star = pd.merge(star_alleles, genes, on=\"star_annotation\").drop_duplicates()\n",
    "\n",
    "# add reference \n",
    "pmid= pmid_table[[\"ref_id\", \"reference_id\" ]]\n",
    "new_star = pd.merge(new_star, pmid, left_on=\"Pubmed_id\",right_on=\"reference_id\").drop_duplicates()\n",
    "\n",
    "# add the other columns \n",
    "last_merg_table_star = star[[\"star_annotation\", \"id\", \"reference_id\", \"drug_id\", \"source\", \"id_in_source\"]]\n",
    "\n",
    "raw_merged_star = pd.merge(new_star, last_merg_table_star, how=\"left\",\n",
    "                      left_on=[\"star_annotation\", \"ref_id\"], right_on = [\"star_annotation\", \"reference_id\"]).drop_duplicates()\n",
    "\n",
    "raw_merged_star = raw_merged_star[[\"id\", \"star_annotation\", \"gene_id\", \"drug_id\", \"phenotype\", \"reference_id_y\",\n",
    "                \"p-value\", \"source\", \"id_in_source\", \"region\", \"Ethnicity\", \"Notes\", \"origin_of_participants\" ]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "id": "2dcb3b6e-43c4-4b60-94a9-d267af88558b",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_merged_star.to_csv(\"../data/new_curated_star.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "b6749590-a8b7-4631-a379-4715ecc6934c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "id                         object\n",
       "star_annotation            object\n",
       "gene_id                    object\n",
       "drug_id                    object\n",
       "allele                     object\n",
       "phenotype                  object\n",
       "reference_id               object\n",
       "P-value                    object\n",
       "source                     object\n",
       "id_in_source                int64\n",
       "region                     object\n",
       "Country_of_Participants    object\n",
       "dtype: object"
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "star.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "id": "1992daa4-d18a-4066-8706-8b5050ed8873",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<pandasgui.gui.PandasGui at 0x7fe8a26890d0>"
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pg.show(raw_merged_star)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "id": "e6de9423-8775-4c49-aa30-dd15846fb54e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>star_annotation</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>drug_id</th>\n",
       "      <th>allele</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>reference_id</th>\n",
       "      <th>P-value</th>\n",
       "      <th>source</th>\n",
       "      <th>id_in_source</th>\n",
       "      <th>region</th>\n",
       "      <th>Country_of_Participants</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>STAR1</td>\n",
       "      <td>CYP1A2*1K</td>\n",
       "      <td>PHARGENE19</td>\n",
       "      <td>DRUG10</td>\n",
       "      <td>*1K</td>\n",
       "      <td>CYP1A2 *1K is associated with decreased metabo...</td>\n",
       "      <td>REF162</td>\n",
       "      <td>&lt; 0.02</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>827698390</td>\n",
       "      <td>Sub-Saharan African</td>\n",
       "      <td>Ethiopia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STAR2</td>\n",
       "      <td>CYP3A4*1</td>\n",
       "      <td>PHARGENE27</td>\n",
       "      <td>DRUG12</td>\n",
       "      <td>*1/*22 + *22/*22</td>\n",
       "      <td>CYP3A4 *1/*22 + *22/*22 is associated with con...</td>\n",
       "      <td>REF152</td>\n",
       "      <td>0.01</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>1184746928</td>\n",
       "      <td>African American/Afro-Caribbean</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STAR3</td>\n",
       "      <td>CYP3A4*22</td>\n",
       "      <td>PHARGENE27</td>\n",
       "      <td>DRUG12</td>\n",
       "      <td>*1/*22 + *22/*22</td>\n",
       "      <td>CYP3A4 *1/*22 + *22/*22 is associated with con...</td>\n",
       "      <td>REF152</td>\n",
       "      <td>0.01</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>1184746928</td>\n",
       "      <td>African American/Afro-Caribbean</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STAR196</td>\n",
       "      <td>CYP3A4*1</td>\n",
       "      <td>PHARGENE27</td>\n",
       "      <td>DRUG55</td>\n",
       "      <td>*1/*1B</td>\n",
       "      <td>CYP3A4 *1/*1B is associated with increased dos...</td>\n",
       "      <td>REF160</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>1449170970</td>\n",
       "      <td>African American/Afro-Caribbean</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>STAR197</td>\n",
       "      <td>CYP3A4*1B</td>\n",
       "      <td>PHARGENE27</td>\n",
       "      <td>DRUG55</td>\n",
       "      <td>*1/*1B</td>\n",
       "      <td>CYP3A4 *1/*1B is associated with increased dos...</td>\n",
       "      <td>REF160</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>1449170970</td>\n",
       "      <td>African American/Afro-Caribbean</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>STAR157</td>\n",
       "      <td>CYP2D6*36</td>\n",
       "      <td>PHARGENE26</td>\n",
       "      <td>DRUG36</td>\n",
       "      <td>*2/*36</td>\n",
       "      <td>CYP2D6 *2/*36 (assigned as extensive metaboliz...</td>\n",
       "      <td>REF76</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>1446897151</td>\n",
       "      <td>African American/Afro-Caribbean</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>STAR202</td>\n",
       "      <td>CYP2D6*2</td>\n",
       "      <td>PHARGENE26</td>\n",
       "      <td>DRUG57</td>\n",
       "      <td>*2/*41</td>\n",
       "      <td>CYP2D6 *2/*41 is associated with increased con...</td>\n",
       "      <td>REF59</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>1449189944</td>\n",
       "      <td>Sub-Saharan African</td>\n",
       "      <td>Canada</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>STAR203</td>\n",
       "      <td>CYP2D6*41</td>\n",
       "      <td>PHARGENE26</td>\n",
       "      <td>DRUG57</td>\n",
       "      <td>*2/*41</td>\n",
       "      <td>CYP2D6 *2/*41 is associated with increased con...</td>\n",
       "      <td>REF59</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>1449189944</td>\n",
       "      <td>Sub-Saharan African</td>\n",
       "      <td>Canada</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>STAR129</td>\n",
       "      <td>CYP2C19*1</td>\n",
       "      <td>PHARGENE23</td>\n",
       "      <td>DRUG40</td>\n",
       "      <td>*1/*17</td>\n",
       "      <td>CYP2C19 *1/*17 is associated with decreased co...</td>\n",
       "      <td>REF161</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>1444828163</td>\n",
       "      <td>Sub-Saharan African</td>\n",
       "      <td>Cameroon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>STAR130</td>\n",
       "      <td>CYP2C19*17</td>\n",
       "      <td>PHARGENE23</td>\n",
       "      <td>DRUG40</td>\n",
       "      <td>*1/*17</td>\n",
       "      <td>CYP2C19 *1/*17 is associated with decreased co...</td>\n",
       "      <td>REF161</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>1444828163</td>\n",
       "      <td>Sub-Saharan African</td>\n",
       "      <td>Cameroon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>203 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          id star_annotation     gene_id drug_id            allele  \\\n",
       "0      STAR1       CYP1A2*1K  PHARGENE19  DRUG10               *1K   \n",
       "1      STAR2        CYP3A4*1  PHARGENE27  DRUG12  *1/*22 + *22/*22   \n",
       "2      STAR3       CYP3A4*22  PHARGENE27  DRUG12  *1/*22 + *22/*22   \n",
       "3    STAR196        CYP3A4*1  PHARGENE27  DRUG55            *1/*1B   \n",
       "4    STAR197       CYP3A4*1B  PHARGENE27  DRUG55            *1/*1B   \n",
       "..       ...             ...         ...     ...               ...   \n",
       "198  STAR157       CYP2D6*36  PHARGENE26  DRUG36            *2/*36   \n",
       "199  STAR202        CYP2D6*2  PHARGENE26  DRUG57            *2/*41   \n",
       "200  STAR203       CYP2D6*41  PHARGENE26  DRUG57            *2/*41   \n",
       "201  STAR129       CYP2C19*1  PHARGENE23  DRUG40            *1/*17   \n",
       "202  STAR130      CYP2C19*17  PHARGENE23  DRUG40            *1/*17   \n",
       "\n",
       "                                             phenotype reference_id P-value  \\\n",
       "0    CYP1A2 *1K is associated with decreased metabo...       REF162  < 0.02   \n",
       "1    CYP3A4 *1/*22 + *22/*22 is associated with con...       REF152    0.01   \n",
       "2    CYP3A4 *1/*22 + *22/*22 is associated with con...       REF152    0.01   \n",
       "3    CYP3A4 *1/*1B is associated with increased dos...       REF160     NaN   \n",
       "4    CYP3A4 *1/*1B is associated with increased dos...       REF160     NaN   \n",
       "..                                                 ...          ...     ...   \n",
       "198  CYP2D6 *2/*36 (assigned as extensive metaboliz...        REF76     NaN   \n",
       "199  CYP2D6 *2/*41 is associated with increased con...        REF59     NaN   \n",
       "200  CYP2D6 *2/*41 is associated with increased con...        REF59     NaN   \n",
       "201  CYP2C19 *1/*17 is associated with decreased co...       REF161     NaN   \n",
       "202  CYP2C19 *1/*17 is associated with decreased co...       REF161     NaN   \n",
       "\n",
       "       source  id_in_source                           region  \\\n",
       "0    PharmGKB     827698390              Sub-Saharan African   \n",
       "1    PharmGKB    1184746928  African American/Afro-Caribbean   \n",
       "2    PharmGKB    1184746928  African American/Afro-Caribbean   \n",
       "3    PharmGKB    1449170970  African American/Afro-Caribbean   \n",
       "4    PharmGKB    1449170970  African American/Afro-Caribbean   \n",
       "..        ...           ...                              ...   \n",
       "198  PharmGKB    1446897151  African American/Afro-Caribbean   \n",
       "199  PharmGKB    1449189944              Sub-Saharan African   \n",
       "200  PharmGKB    1449189944              Sub-Saharan African   \n",
       "201  PharmGKB    1444828163              Sub-Saharan African   \n",
       "202  PharmGKB    1444828163              Sub-Saharan African   \n",
       "\n",
       "    Country_of_Participants  \n",
       "0                  Ethiopia  \n",
       "1                       USA  \n",
       "2                       USA  \n",
       "3                       USA  \n",
       "4                       USA  \n",
       "..                      ...  \n",
       "198                     USA  \n",
       "199                  Canada  \n",
       "200                  Canada  \n",
       "201                Cameroon  \n",
       "202                Cameroon  \n",
       "\n",
       "[203 rows x 12 columns]"
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "star"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "763e401a-080e-479c-9465-f17c26d17899",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
